src/relations/async_replication.py

```diff
@@ -0,0 +1,665 @@
+            self.charm.update_config()
+            if self._is_primary_cluster() and self.charm.unit.is_leader():
+                self._update_primary_cluster_data()
+                # If this is a standby cluster, remove the information from DCS to make it
+                # a normal cluster.
+                if self.charm._patroni.get_standby_leader() is not None:
+                    self.charm._patroni.promote_standby_cluster()
+                    try:
+                        for attempt in Retrying(stop=stop_after_delay(60), wait=wait_fixed(3)):
+                            with attempt:
+                                if not self.charm.is_primary:
+                                    raise ClusterNotPromotedError()
+                    except RetryError:
+                        logger.debug(
+                            "Deferring on_async_relation_changed: standby cluster not promoted yet."
+                        )
+                        event.defer()
+                        return True
+            self.charm._peers.data[self.charm.unit].update({
+                "unit-promoted-cluster-counter": self._get_highest_promoted_cluster_counter_value()
+            })
+            self.charm._set_active_status()
+            return True
+        return False
+
+    def _configure_standby_cluster(self, event: RelationChangedEvent) -> bool:
+        """Configure the standby cluster."""
+        relation = self._relation
+        if relation.name == ASYNC_REPLICA_RELATION:
+            # Update the secrets between the clusters.
+            primary_cluster_info = relation.data[relation.app].get("primary-cluster-data")
+            secret_id = (
+                None
+                if primary_cluster_info is None
+                else json.loads(primary_cluster_info).get("secret-id")
+            )
+            try:
+                secret = self.charm.model.get_secret(id=secret_id, label=self._secret_label)
+            except SecretNotFoundError:
+                logger.debug("Secret not found, deferring event")
+                event.defer()
+                return False
+            credentials = secret.peek_content()
+            for key, password in credentials.items():
+                user = key.split("-password")[0]
+                self.charm.set_secret(APP_SCOPE, key, password)
+                logger.debug("Synced %s password", user)
+        system_identifier, error = self.get_system_identifier()
+        if error is not None:
+            raise Exception(error)
+        if system_identifier != relation.data[relation.app].get("system-id"):
+            # Store current data in a tar.gz file.
+            logger.info("Creating backup of pgdata folder")
+            filename = f"{POSTGRESQL_DATA_PATH}-{str(datetime.now()).replace(' ', '-').replace(':', '-')}.tar.gz"
+            self.container.exec(
+                f"tar -zcf {filename} {POSTGRESQL_DATA_PATH}".split()
+            ).wait_output()
```

Comment:
Please create followup tickets:
* pack archive in background (otherwise promotion to Standby will take a LOT of time if local DB is huge)
* warn users about available backups to clean (free disk space topic), goss is a good match here

Reply:
Sure: https://warthogs.atlassian.net/browse/DPE-4256